### PR TITLE
Add Liquidity Pools methods

### PR DESCRIPTION
**Pull request by onnovisser**
### Description

Adds a Vault entity with methods to query investments and place invest/redeem orders.

#17

### CODE REVIEW

**Review by sophialittlejohn — APPROVED** [review]

File: src/Vault.ts [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+   * @param investor - The address of the investor
+   */
+  investment(investor: string) {
+    const address = investor.toLowerCase() as HexString
+    return this._query(['investment', address], () =>
+      combineLatest([
+        this.investmentCurrency(),
+        this.shareCurrency(),
+        this.network._investmentManager(),
+        this._restrictionManager(),
```
Comment (hieronx) [inline src/Vault.ts:133]: Will this be somehow magically combined into 1 big multicall, or will it be 4 separate calls?

If the latter, would it be worth combining more of these read calls into 1 method?

File: src/Vault.ts [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+            defer(async () => {
+              const client = this._root.getClient(this.chainId)!
+              const vault = getContract({ address: this.address, abi: ABI.LiquidityPool, client })
+              const investmentManager = getContract({
+                address: investmentManagerAddress,
+                abi: ABI.InvestmentManager,
+                client,
+              })
+
+              const [isAllowedToInvest, maxDeposit, maxRedeem, investment] = await Promise.all([
```
Comment (hieronx) [inline src/Vault.ts:149]: I assume this will be multicall, right?

File: src/Vault.ts [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+                pendingRedeemShares: new Token(pendingRedeem, shareCurrency.decimals),
+                claimableCancelInvestCurrency: new Currency(claimableCancelInvestCurrency, investmentCurrency.decimals),
+                claimableCancelRedeemShares: new Token(claimableCancelRedeemShares, shareCurrency.decimals),
+                hasPendingCancelInvestRequest,
+                hasPendingCancelRedeemRequest,
+                investmentCurrency,
+                shareCurrency,
+              }
+            }).pipe(
+              repeatOnEvents(
```
Comment (hieronx) [inline src/Vault.ts:186]: Nice :)

File: src/Vault.ts [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+   */
+  cancelInvestOrder() {
+    const self = this
+    return this._transactSequence(async function* ({ walletClient, signingAddress, publicClient }) {
+      const { centrifugeRouter } = lpConfig[self.chainId]!
+      const [estimate, investment] = await Promise.all([self.network._estimate(), self.investment(signingAddress)])
+
+      if (investment.pendingInvestCurrency.isZero()) throw new Error('No order to cancel')
+
+      yield* doTransaction('Cancel Invest Order', publicClient, () =>
```
Comment (hieronx) [inline src/Vault.ts:326]: Will these tx descriptions be passed through and shown in the UI? If so, I would stick to our pattern of not capitalizing:
```suggestion
      yield* doTransaction('Cancel invest order', publicClient, () =>
```

File: src/Vault.ts [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+        }))
+      )
+    )
+  }
+
+  /**
+   * Place an order to invest funds in the vault. If an order exists, it will increase the amount.
+   * @param investAmount - The amount to invest in the vault
+   */
+  increaseInvestOrder(investAmount: bigint | number) {
```
Comment (hieronx) [inline src/Vault.ts:237]: Nit: param name here is `investAmount`, while for `increaseRedeemOrder` it is `shares`. 

Would either use `assets` here, or `redeemAmount` in `increaseRedeemOrder`

File: src/Vault.ts [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+        })
+      )
+    }, this.chainId)
+  }
+
+  /**
+   * Claim any outstanding fund shares after an investment has gone through, or funds after an redemption has gone through.
+   * @param receiver - The address that should receive the funds. If not provided, the investor's address is used.
+   */
+  claim(receiver?: string) {
```
Comment (hieronx) [inline src/Vault.ts:391]: Would be good to also add an optional `controller` param here, so we can actually use the SDK to build a bot to claim for others

File: src/Vault.test.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
+import { expect } from 'chai'
+import { firstValueFrom, lastValueFrom, skip, skipWhile, tap, toArray } from 'rxjs'
+import sinon from 'sinon'
+import { ABI } from './abi/index.js'
+import { Pool } from './Pool.js'
+import { PoolNetwork } from './PoolNetwork.js'
+import { context } from './tests/setup.js'
+import { Vault } from './Vault.js'
+
+const poolId = '2779829532'
```
Comment (hieronx) [inline src/Vault.test.ts:10]: Really love these impersonated fork tests!!

File: src/Vault.ts (hunk 1 above)
Comment (onnovisser) [reply src/Vault.ts:134]: Viem has the option to automatically batch these, which is awesome, so these will be done as a multicall 🥳

Comment (hieronx) [reply src/Vault.ts:134]: That is wild! 🔥

**Pull request merged**